Simple quality filtering for FASTQ files
One common task when working with FASTQ fikes is taking a large set of sequencing reads and filtering them based on their quality scores. We are going to read FASTQ data, and filter it to pick out only those records whose PHRED quality scores are all above some threshold.

In [20]:
!wget ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR020/SRR020192/SRR020192.fastq.gz 

'wget' is not recognized as an internal or external command,
operable program or batch file.


In [22]:
import gzip
import shutil

compressed_filename = "SRR020192.fastq.gz"
uncompressed_filename = "SRR020192.fastq"

with gzip.open(compressed_filename, 'rb') as f_in:
    with open(uncompressed_filename, 'wb') as f_out:
        shutil.copyfileobj(f_in, f_out)

print("File decompressed successfully.")


File decompressed successfully.


In [23]:
from Bio import SeqIO
count = 0
for rec in SeqIO.parse("SRR020192.fastq", "fastq"):
    count += 1
print(f"We have {count} reads.")

We have 41892 reads.


In [24]:
# how phred quality scores look like
rec = SeqIO.parse("SRR020192.fastq", "fastq")
rec = next(rec)
rec.letter_annotations["phred_quality"]

[24,
 23,
 27,
 30,
 30,
 30,
 23,
 23,
 24,
 23,
 23,
 30,
 28,
 27,
 25,
 25,
 27,
 27,
 27,
 22,
 22,
 24,
 18,
 18,
 18,
 30,
 19,
 19,
 23,
 23,
 30,
 30,
 32,
 32,
 32,
 30,
 24,
 23,
 23,
 27,
 30,
 32,
 30,
 32,
 29,
 28,
 28,
 17,
 17,
 17,
 17,
 24,
 17,
 17,
 13,
 15,
 17,
 25,
 25,
 24,
 24,
 23,
 27,
 27,
 15,
 15,
 15,
 15,
 15,
 17,
 17,
 11,
 15,
 15]

In [25]:
good_reads = (
    rec
    for rec in SeqIO.parse("SRR020192.fastq", "fastq")
    if min(rec.letter_annotations["phred_quality"]) >= 20
)
count = SeqIO.write(good_reads, "good_quality.fastq", "fastq")
print(f"Saved {count} reads.")

Saved 20050 reads.


This pulled out only 20050 reads out of the 41892 present